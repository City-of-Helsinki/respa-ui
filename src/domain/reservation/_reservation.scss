@import 'manage/page/manageReservationsPage';
@import 'manage/list/manageReservationsList';
@import 'manage/status/manageReservationsStatus';
