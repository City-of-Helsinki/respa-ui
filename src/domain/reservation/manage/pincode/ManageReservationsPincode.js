import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { Panel } from 'react-bootstrap';

import { isAccessCodeGenerated, isAccessCodePending } from '../../../../../app/shared/reservation-access-code/helpers';
import TooltipOverlay from '../../../../common/tooltip/TooltipOverlay';
import iconClock from '../../../../../app/assets/icons/clock-o.svg';

class ManageReservationsPincode extends Component {
  renderPincodeField() {
    const { reservation } = this.props;
    if (isAccessCodeGenerated(reservation)) {
      return (
        <TooltipOverlay
          content={(
            <Panel>
              <Panel.Body>
                <Panel.Title><FormattedMessage id="common.accessCodeLabel" /></Panel.Title>
                <p>{reservation.accessCode}</p>
              </Panel.Body>
            </Panel>
          )}
        >
          <span>****</span>
        </TooltipOverlay>
      );
    }
    if (isAccessCodePending(reservation, reservation.resource)) {
      return (
        <TooltipOverlay
          content={<FormattedMessage id="ReservationAccessCode.pending" />}
        >
          <img alt="reservationAccessCodePending" src={iconClock} />
        </TooltipOverlay>
      );
    }

    return '';
  }

  render() {
    return (
      <div className="app-ManageReservationPincode">
        {this.renderPincodeField()}
      </div>
    );
  }
}

ManageReservationsPincode.propTypes = {
  reservation: PropTypes.object.isRequired
};

export default ManageReservationsPincode;
