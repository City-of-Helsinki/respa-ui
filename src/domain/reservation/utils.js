import moment from 'moment';
import find from 'lodash/find';
import sortBy from 'lodash/sortBy';
import clone from 'lodash/clone';
import tail from 'lodash/tail';
import last from 'lodash/last';

export const combineReservations = (reservations) => {
  if (!reservations || !reservations.length) {
    return [];
  }

  const sorted = sortBy(reservations, 'begin');
  const initialValue = [clone(sorted[0])];

  return tail(sorted).reduce((previous, current) => {
    if (current.begin === last(previous).end) {
      last(previous).end = current.end;
    } else {
      previous.push(clone(current));
    }
    return previous;
  }, initialValue);
};

export const getNextAvailableTime = (reservations, fromMoment = moment()) => {
  const combinedReservations = combineReservations(reservations);
  if (!combinedReservations.length || fromMoment < moment(combinedReservations[0].begin)) {
    return fromMoment;
  }

  const ongoingReservation = find(combinedReservations, reservation => (
    moment(reservation.begin) <= fromMoment && fromMoment < moment(reservation.end)
  ));

  return ongoingReservation ? moment(ongoingReservation.end) : fromMoment;
};

export const getCurrentReservation = (reservations) => {
  const now = moment();
  return find(reservations, (reservation) => {
    return moment(reservation.begin) < now && now < moment(reservation.end);
  });
};
