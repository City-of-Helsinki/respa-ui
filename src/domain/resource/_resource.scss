@import './card/resourceCard';
@import './card/slider/resourceCardSlider';
