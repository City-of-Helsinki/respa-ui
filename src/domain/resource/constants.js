export const resourceProductTypes = Object.freeze({
  RENT: 'rent',
  EXTRA: 'extra',
});

export const resourcePriceTypes = Object.freeze({
  HOURLY: 'hourly',
  DAILY: 'daily',
  WEEKLY: 'weekly',
  FIXED: 'fixed',
});

export const productPriceType = Object.freeze({
  PER_PERIOD: 'per_period',
  FIXED: 'fixed',
});

export const resourcePeriodLengthType = Object.freeze({
  WITHIN_DAY: 'within_day',
  WHOLE_DAY: 'whole_day',
  OVERNIGHT: 'over_night',
});
