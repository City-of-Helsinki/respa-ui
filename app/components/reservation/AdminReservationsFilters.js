import map from 'lodash/collection/map';
import sortBy from 'lodash/collection/sortBy';
import React, { Component, PropTypes } from 'react';
import Select from 'react-select';

import { RESERVATION_STATE_LABELS } from 'constants/AppConstants';

let stateOptions = map(RESERVATION_STATE_LABELS, (value, key) => {
  return {
    label: value.labelText,
    value: key,
  };
});
stateOptions = sortBy(stateOptions, 'label');
stateOptions.unshift({ label: 'Kaikki', value: 'all' });

class AdminReservationsFilters extends Component {
  render() {
    const {
      onFiltersChange,
      filters,
    } = this.props;

    return (
      <div className="reservations-filters">
        <h4>Varauksen status</h4>
        <Select
          className="reservation-state-select"
          clearable={false}
          name="reservation-state-select"
          onChange={(value) => onFiltersChange({ state: value })}
          options={stateOptions}
          value={filters.state}
        />
      </div>
    );
  }
}

AdminReservationsFilters.propTypes = {
  filters: PropTypes.object.isRequired,
  onFiltersChange: PropTypes.func.isRequired,
};

export default AdminReservationsFilters;
