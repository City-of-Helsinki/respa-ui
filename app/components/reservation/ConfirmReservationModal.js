import pick from 'lodash/object/pick';
import camelCase from 'lodash/string/camelCase';
import React, { Component, PropTypes } from 'react';
import Modal from 'react-bootstrap/lib/Modal';

import CompactReservationsList from 'components/common/CompactReservationsList';
import { RESERVATION_FORM_FIELDS } from 'constants/AppConstants';
import ReservationForm from 'containers/ReservationForm';

class ConfirmReservationModal extends Component {
  constructor(props) {
    super(props);
    this.getFormFields = this.getFormFields.bind(this);
    this.getFormInitialValues = this.getFormInitialValues.bind(this);
    this.onConfirm = this.onConfirm.bind(this);
    this.renderIntroTexts = this.renderIntroTexts.bind(this);
  }

  onConfirm(values) {
    const { onClose, onConfirm } = this.props;
    onClose();
    onConfirm(values);
  }

  getFormFields() {
    const { resource } = this.props;
    const isAdmin = resource.userPermissions.isAdmin;
    const formFields = [];
    if (resource.needManualConfirmation) {
      formFields.push(...RESERVATION_FORM_FIELDS);
    }

    if (isAdmin) {
      formFields.push('comments');
      if (resource.needManualConfirmation) {
        formFields.push('staffEvent');
      }
    }

    return formFields;
  }

  getFormInitialValues() {
    const {
      isEditing,
      reservationsToEdit,
      selectedReservations,
    } = this.props;
    let reservation;

    if (isEditing) {
      reservation = reservationsToEdit.length ? reservationsToEdit[0] : null;
    } else {
      reservation = selectedReservations.length ? selectedReservations[0] : null;
    }

    return reservation ? pick(reservation, ['comments', ...RESERVATION_FORM_FIELDS]) : {};
  }

  getModalTitle(isEditing, isPreliminaryReservation) {
    if (isEditing) {
      return 'Muutosten vahvistus';
    }
    if (isPreliminaryReservation) {
      return 'Alustava varaus';
    }
    return 'Varauksen vahvistus';
  }

  renderIntroTexts() {
    const {
      isEditing,
      isPreliminaryReservation,
      reservationsToEdit,
      selectedReservations,
    } = this.props;

    if (isEditing) {
      return (
        <div>
          <p><strong>Oletko varma että haluat muuttaa varaustasi?</strong></p>
          <p>Ennen muutoksia:</p>
          <CompactReservationsList reservations={reservationsToEdit} />
          <p>Muutosten jälkeen:</p>
          <CompactReservationsList reservations={selectedReservations} />
        </div>
      );
    }

    let helpText;

    if (isPreliminaryReservation) {
      helpText = selectedReservations.length === 1 ?
        'Olet tekemässä alustavaa varausta ajalle:' :
        'Olet tekemässä alustavaa varausta ajoille:';
    } else {
      helpText = selectedReservations.length === 1 ?
        'Oletko varma että haluat tehdä varauksen ajalle:' :
        'Oletko varma että haluat tehdä varaukset ajoille:';
    }

    return (
      <div>
        <p><strong>{helpText}</strong></p>
        <CompactReservationsList reservations={selectedReservations} />
        {isPreliminaryReservation && (
          <p>
            Täytä vielä seuraavat tiedot alustavaa varausta varten.
            Tähdellä (*) merkityt tiedot ovat pakollisia.
          </p>
        )}
      </div>
    );
  }

  render() {
    const {
      isEditing,
      isMakingReservations,
      isPreliminaryReservation,
      onClose,
      resource,
      show,
    } = this.props;

    const requiredFormFields = resource.requiredReservationExtraFields.map((field) => {
      return camelCase(field);
    });

    return (
      <Modal
        animation={false}
        className="confirm-reservation-modal"
        onHide={onClose}
        show={show}
      >
        <Modal.Header closeButton>
          <Modal.Title>
            {this.getModalTitle(isEditing, isPreliminaryReservation)}
          </Modal.Title>
        </Modal.Header>

        <Modal.Body>
          {this.renderIntroTexts()}
          <ReservationForm
            fields={this.getFormFields()}
            initialValues={this.getFormInitialValues()}
            isMakingReservations={isMakingReservations}
            onClose={onClose}
            onConfirm={this.onConfirm}
            requiredFields={requiredFormFields}
          />
        </Modal.Body>
      </Modal>
    );
  }
}

ConfirmReservationModal.propTypes = {
  isEditing: PropTypes.bool.isRequired,
  isMakingReservations: PropTypes.bool.isRequired,
  isPreliminaryReservation: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  onConfirm: PropTypes.func.isRequired,
  reservationsToEdit: PropTypes.array.isRequired,
  resource: PropTypes.object.isRequired,
  selectedReservations: PropTypes.array.isRequired,
  show: PropTypes.bool.isRequired,
};

export default ConfirmReservationModal;
