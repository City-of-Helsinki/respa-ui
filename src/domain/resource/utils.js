import round from 'lodash/round';
import filter from 'lodash/filter';
import findIndex from 'lodash/findIndex';
import get from 'lodash/get';
import find from 'lodash/find';
import moment from 'moment';

import * as urlUtils from '../../common/url/utils';
import * as dataUtils from '../../common/data/utils';
import * as reservationUtils from '../reservation/utils';
import constants from '../../../app/constants/AppConstants';
import forEach from "lodash/forEach";


/**
 * Getter for resource page link.
 * Returns a resource page url with the given query params.
 * @param resource
 * @param query
 * @returns {string}
 */
export const getResourcePageLink = (resource, query) => {
  return urlUtils.getLinkString(`/resources/${resource.id}`, query);
};

/**
 * Getter for formatted unit address string.
 * @param unit
 * @param locale
 * @returns {string}
 */
export const getUnitAddress = (unit, locale) => {
  let postal;
  const streetAddress = dataUtils.getLocalizedFieldValue(unit.street_address, locale);
  if (unit && unit.address_zip) {
    postal = unit.address_zip;
  }
  if (unit && unit.municipality) {
    postal = postal ? `${postal} ${unit.municipality}` : unit.municipality;
  }

  return `${streetAddress}${postal ? ` ${postal}` : ''}`;
};

/**
 * Getter for formatted resource distance string.
 * @param resource
 * @returns {string}
 */
export const getResourceDistance = (resource) => {
  const km = resource.distance / 1000;
  return km ? `${round(km, km < 10 ? 1 : null)} km` : '';
};

/**
 * Getter for price string used in resource cards.
 * @param resource
 * @param t
 * @returns {string|*}
 */
export const getPrice = (resource, t) => {
  const minPricePerHour = resource.min_price_per_hour;
  const maxPricePerHour = resource.max_price_per_hour;

  if (!(minPricePerHour || maxPricePerHour)) {
    return t('ResourceIcons.free');
  }

  if (minPricePerHour && maxPricePerHour && minPricePerHour !== maxPricePerHour) {
    return `${Number(minPricePerHour)} - ${Number(maxPricePerHour)} €/h`;
  }

  const priceString = maxPricePerHour || minPricePerHour;
  const price = priceString !== 0 ? Number(priceString) : 0;
  if (price === 0) {
    return t('ResourceIcons.free');
  }

  return price ? `${price} €/h` : null;
};

export const isFree = (resource) => {
  const minPricePerHour = resource.min_price_per_hour;
  const maxPricePerHour = resource.max_price_per_hour;

  if (!(minPricePerHour || maxPricePerHour)) {
    return true;
  }

  if (minPricePerHour && maxPricePerHour && minPricePerHour !== maxPricePerHour) {
    return false;
  }

  const priceString = maxPricePerHour || minPricePerHour;
  const price = priceString !== 0 ? Number(priceString) : 0;

  return price === 0;
};

export const getOpeningHours = (resource, date = null) => {
  if (!resource) {
    return null;
  }

  const openingHours = get(resource, 'opening_hours', []);
  if (!openingHours.length) {
    return null;
  }

  const index = date ? findIndex(openingHours, item => item.date === date) : 0;
  return {
    closes: get(openingHours, `[${index}].closes`),
    opens: get(openingHours, `[${index}].opens`),
  };
};

const reservingIsRestricted = (resource, date) => {
  if (!date) {
    return false;
  }

  const isAdmin = get(resource, 'user_permissions.is_admin', false);
  const isLimited = resource.reservable_before && moment(resource.reservable_before).isBefore(moment(date), 'day');

  return isLimited && !isAdmin;
};

export const getOpenReservations = (resource) => {
  return filter(resource.reservations, (reservation) => {
    return reservation.state !== 'cancelled' && reservation.state !== 'denied';
  });
};

export const getAvailabilityDataForNow = (resource, date) => {
  const openingHours = getOpeningHours(resource, date);
  const reservations = getOpenReservations(resource);

  if (!openingHours || !openingHours.closes || !openingHours.opens) {
    return { status: 'closed', bsStyle: 'danger' };
  }

  const nowMoment = moment();
  const opensMoment = moment(openingHours.opens);
  const closesMoment = moment(openingHours.closes);
  const beginMoment = nowMoment > opensMoment ? nowMoment : opensMoment;

  if (nowMoment > closesMoment) {
    return { status: 'closed', bsStyle: 'danger' };
  }

  const currentReservation = reservationUtils.getCurrentReservation(reservations);
  if (currentReservation || nowMoment < opensMoment) {
    const nextAvailableTime = reservationUtils.getNextAvailableTime(reservations, beginMoment);
    if (nextAvailableTime < closesMoment) {
      return {
        status: 'availableAt',
        bsStyle: 'danger',
        values: { time: nextAvailableTime.format(constants.TIME_FORMAT) },
      };
    }
    return { status: 'reserved', bsStyle: 'danger' };
  }

  return { status: 'available', bsStyle: 'success' };
};

export const getAvailabilityDataForWholeDay = (resource, date) => {
  const openingHours = getOpeningHours(resource, date);

  if (!openingHours || !openingHours.closes || !openingHours.opens) {
    return { status: 'closed', bsStyle: 'danger' };
  }

  if (reservingIsRestricted(resource, date)) {
    return { status: 'reservingRestricted', bsStyle: 'danger' };
  }

  const reservations = getOpenReservations(resource);

  const opensMoment = moment(openingHours.opens);
  const closesMoment = moment(openingHours.closes);
  let total = closesMoment - opensMoment;

  forEach(reservations, (reservation) => {
    const resBeginMoment = moment(reservation.begin);
    if (!resBeginMoment.isSame(opensMoment, 'd')) {
      return;
    }

    const resEndMoment = moment(reservation.end);
    total -= resEndMoment;
    total += resBeginMoment;
  });

  const asHours = moment.duration(total).asHours();
  const rounded = Math.ceil(asHours * 2) / 2;

  if (rounded === 0) {
    return { status: 'reserved', bsStyle: 'danger' };
  }

  return {
    status: 'availableTime',
    bsStyle: 'success',
    values: { hours: rounded },
  };
};
