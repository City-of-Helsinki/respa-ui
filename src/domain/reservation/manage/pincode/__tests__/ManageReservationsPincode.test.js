import React from 'react';
import toJSON from 'enzyme-to-json';
import { shallow } from 'enzyme';

import ManageReservationsPincode from '../ManageReservationsPincode';
import reservation from '../../../../../common/data/fixtures/reservation';
import resource from '../../../../../common/data/fixtures/resource';

describe('ManageReservationsPincode', () => {
  function getWrapper(props) {
    return shallow(
      <ManageReservationsPincode {...props} />
    );
  }

  const pendingCodeResource = resource.attr('generateAccessCodes', false);

  test('renders correctly', () => {
    const wrapper = getWrapper({ reservation: reservation.build() });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test('show Icon when accessCode pending', () => {
    const wrapper = getWrapper({
      reservation: reservation.build({ resource: pendingCodeResource.build() })
    });
    expect(wrapper.debug()).toContain('img');
  });

  test('show **** when accessCode exist', () => {
    reservation.attr('accessCode', 1234);
    const wrapper = getWrapper({ reservation: reservation.build() });
    expect(wrapper.html()).toContain('****');
  });
});
