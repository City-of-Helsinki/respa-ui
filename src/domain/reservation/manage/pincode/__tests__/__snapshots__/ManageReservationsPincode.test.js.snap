// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageReservationPincode renders correctly 1`] = `
<div
  className="app-ManageReservationPincode"
/>
`;
